src/Protocols/Wishbone.hs

```diff
@@ -0,0 +1,184 @@
+{-|
+Types modelling the Wishbone bus protocol.
+-}
+
+{-# LANGUAGE FlexibleContexts     #-}
+{-# LANGUAGE PatternSynonyms      #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -fconstraint-solver-iterations=0 #-}
+
+module Protocols.Wishbone where
+
+import           Clash.Prelude         (DivRU, Nat, Type, (:::))
+import qualified Clash.Prelude         as C
+
+import           Clash.Signal.Internal (Signal (..))
+
+import           Control.DeepSeq       (NFData)
+import qualified Data.Bifunctor        as B
+import           Prelude               hiding (head, not, (&&))
+import           Protocols
+import           Protocols.Internal
+
+-- | Data communicated from a Wishbone Master to a Wishbone Slave
+data WishboneM2S addressWidth selWidth dat
+  = WishboneM2S
+  { -- | ADR
+    addr                :: "ADR" ::: C.BitVector addressWidth
+    -- | DAT
+  , writeData           :: "DAT_MOSI" ::: dat
+    -- | SEL
+  , busSelect           :: "SEL" ::: C.BitVector selWidth
+    -- | LOCK
+  , lock                :: "LOCK" ::: Bool
+    -- | CYC
+  , busCycle            :: "CYC" ::: Bool
+    -- | STB
+  , strobe              :: "STB" ::: Bool
+    -- | WE
+  , writeEnable         :: "WE" ::: Bool
+    -- | CTI
+  , cycleTypeIdentifier :: "CTI" ::: CycleTypeIdentifier
+    -- | BTE
+  , burstTypeExtension  :: "BTE" ::: BurstTypeExtension
```

Comment:
I'd copy+paste the documentation for these field from the Wishbone spec

Reply:
Done that! I left the original signal names from the spec (for example `STB_O`) for potential better searchability.